### PR TITLE
Separate vchan to read and write

### PR DESCRIPTION
**Pull request by MykolaSolyanko**

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -39,17 +39,25 @@ type Downloader struct {
...
-	XSPath             string     `json:"xsPath"`
-	Domain             int        `json:"domain"`
-	IAMPublicServerURL string     `json:"iamPublicServerUrl"`
-	CMServerURL        string     `json:"cmServerUrl"`
-	CertStorage        string     `json:"certStorage"`
-	CACert             string     `json:"caCert"`
-	ImageStoreDir      string     `json:"imageStoreDir"`
-	Downloader         Downloader `json:"downloader"`
+	WorkingDir string `json:"workingDir"`
+	// XSPath             string     `json:"xsPath"`
```
Comment (al1img) [inline config/config.go:51]: Remove comments

File: config/config.go [hunk 2]
```diff
@@ -39,17 +39,25 @@ type Downloader struct {
 	MaxRetryDelay          aostypes.Duration `json:"maxRetryDelay"`
 }
 
+// VchanConfig vchan configuration.
+type VchanConfig struct {
+	Domain int    `json:"domain"`
```
Comment (al1img) [inline config/config.go:44]: Domain should be the same. So, what about having one common structure?
```json
"vchanWrite": {
		"xsRxPath": "xsPathWrite",
                "xsTxPath": "xsPathWrite",
		"domain": 2
	},
```

Comment (MykolaSolyanko) [reply config/config.go]: Agree, this solution is much better

File: vchanmanager/vchan.go [hunk 3]
```diff
@@ -40,6 +39,17 @@ import (
...
+  if (ctrl == NULL) {
+    return NULL;
+  }
+
+  ctrl->blocking = 1;
+
+  return ctrl;
+}
+
 static void fdZero(void *set) {
```
Comment (al1img) [inline vchanmanager/vchan.go:53]: FD macro still used?

Comment (MykolaSolyanko) [reply vchanmanager/vchan.go]: removed

File: config/config.go [hunk 4]
```diff
@@ -39,17 +39,23 @@ type Downloader struct {
 	MaxRetryDelay          aostypes.Duration `json:"maxRetryDelay"`
 }
 
+// VchanConfig vchan configuration.
+type VchanConfig struct {
```
Comment (Eabramov84) [inline config/config.go:43]: VChanConfig

File: vchanmanager/vchanmanager.go [hunk 5]
```diff
@@ -59,22 +59,24 @@ type Unpacker interface {
 // VChanItf interface for vchan.
 type VChanItf interface {
 	Init(domain int, xsPath string) error
-	Read(ctx context.Context) ([]byte, error)
+	Read() ([]byte, error)
```
Comment (Eabramov84) [inline vchanmanager/vchanmanager.go:62]: maybe use the next interface to read and write.
	Read(b []byte) (n int, err error)
	Write(b []byte) (n int, err error)

File: vchanmanager/vchanmanager.go [hunk 6]
```diff
@@ -59,22 +59,24 @@ type Unpacker interface {
...
-	downloadManager      Downloader
-	unpackerManager      Unpacker
+	sync.Mutex
+
+	vchanReader VChanItf
+	vchanWriter VChanItf
+	recvChan    chan []byte
+	sendChan    chan []byte
+	cancel      context.CancelFunc
+	// connectionLostNotify chan struct{}
```
Comment (Eabramov84) [inline vchanmanager/vchanmanager.go:76]: remove commented code

File: vchanmanager/vchan.go [hunk 7]
```diff
@@ -138,83 +142,29 @@ func (v *VChan) Write(data []byte) (err error) {
...
-			if err != nil {
-				return nil, err
-			}
+func (v *VChan) readVchan(buffSize C.size_t) ([]byte, error) {
+	buffer := make([]byte, buffSize)
 
-			copy(buffer[bytesRead:], readBuff)
-			bytesRead += C.size_t(len(readBuff))
-		}
+	n, errno := C.libxenvchan_read(v.vchan, unsafe.Pointer(&buffer[0]), buffSize)
```
Comment (al1img) [inline vchanmanager/vchan.go:148]: Should we use `libxenvchan_recv` instead? Looks like it is more suitable for blocking API.

File: vchanmanager/vchan.go [hunk 8]
```diff
@@ -138,83 +142,29 @@ func (v *VChan) Write(data []byte) (err error) {
...
-	for n < C.int(buffSize) {
-		v.Lock()
-		ret, errno := C.libxenvchan_write(v.vchan, unsafe.Pointer(&buffer[n]), buffSize-C.size_t(n))
-		v.Unlock()
-		if ret < 0 {
-			return aoserrors.Errorf("vchan server write failed: %v", errno)
-		}
-
-		n += ret
+	n, errno := C.libxenvchan_write(v.vchan, unsafe.Pointer(&buffer[0]), C.size_t(len(buffer)))
```
Comment (al1img) [inline vchanmanager/vchan.go:161]: Shall we use `libxenvchan_send`?

File: vchanmanager/vchanmanager_test.go [hunk 9]
```diff
@@ -87,11 +87,14 @@ func TestReadWriteData(t *testing.T) {
...
-	}, &testDownloader{}, nil, tVchan)
+		Vchan: config.VchanConfig{
+			XsRxPath: "/tmp/xs_rx",
+			XsTxPath: "/tmp/xs_tx",
+			Domain:   1,
+		},
+	}, &testDownloader{}, nil, tVchan, tVchan)
 	if err != nil {
-		t.Errorf("Can't create a new communication manager: %v", err)
+		t.Errorf("Can't create a new vcha manager: %v", err)
```
Comment (al1img) [inline vchanmanager/vchanmanager_test.go:97]: vchannel

File: vchanmanager/vchan.go (hunk 7 above)
Comment (MykolaSolyanko) [reply vchanmanager/vchan.go:148]: Agree. Done

File: vchanmanager/vchan.go (hunk 8 above)
Comment (MykolaSolyanko) [reply vchanmanager/vchan.go:161]: Agree. Done

File: vchanmanager/vchanmanager_test.go (hunk 9 above)
Comment (MykolaSolyanko) [reply vchanmanager/vchanmanager_test.go:97]: Done

File: config/config.go [hunk 10]
```diff
@@ -39,17 +39,23 @@ type Downloader struct {
...
-	XSPath             string     `json:"xsPath"`
-	Domain             int        `json:"domain"`
-	IAMPublicServerURL string     `json:"iamPublicServerUrl"`
-	CMServerURL        string     `json:"cmServerUrl"`
-	CertStorage        string     `json:"certStorage"`
-	CACert             string     `json:"caCert"`
-	ImageStoreDir      string     `json:"imageStoreDir"`
-	Downloader         Downloader `json:"downloader"`
+	WorkingDir         string      `json:"workingDir"`
+	Vchan              VchanConfig `json:"vchan"`
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply config/config.go:52]: Done

File: vchanmanager/vchanmanager.go (hunk 6 above)
Comment (MykolaSolyanko) [reply vchanmanager/vchanmanager.go:76]: Done

File: config/config.go (hunk 4 above)
Comment (MykolaSolyanko) [reply config/config.go:43]: Done

**Review by al1img — APPROVED** [review]

**Review by Eabramov84 — APPROVED** [review]

**Pull request merged**